File: python/problem_40.py
from typing import List


def single_number(nums: List[int]) -> int:
    result = 0
    for bit in range(64):
        total = sum((n >> bit) & 1 for n in nums)
        if total % 3 != 0:
            result |= (1 << bit)
    return result


def main():
    print("Problem 40:")
    assert single_number([6, 1, 3, 3, 3, 6, 6]) == 1
    assert single_number([13, 19, 13, 13])       == 19
    assert single_number([0, 0, 0, 5])           == 5
    print("All Tests Passed")

File: rust/problem_40.rs
pub fn single_number(nums: &[i64]) -> i64 {
    let mut result = 0i64;
    for bit in 0..64 {
        let total: i64 = nums.iter().map(|&n| (n >> bit) & 1).sum();
        if total % 3 != 0 { result |= 1 << bit; }
    }
    result
}

pub fn problem40() {
    println!("Problem 40:");
    assert_eq!(single_number(&[6,1,3,3,3,6,6]),  1);
    assert_eq!(single_number(&[13,19,13,13]),    19);
    assert_eq!(single_number(&[0,0,0,5]),         5);
    println!("All Tests Passed");
}
